crates/biome_deserialize/src/lib.rs

```diff
@@ -1,31 +1,346 @@
+    ) -> Option<Self::Output> {
+        debug_assert!(
+            !Self::EXPECTED_TYPE.contains(ExpectedType::NUMBER),
+            "This method should be implemented because the type is expected."
+        );
+        diagnostics.push(DeserializationDiagnostic::new_incorrect_type(
+            Self::EXPECTED_TYPE,
+            range,
+        ));
+        None
+    }
+
+    /// The visited value is a `string`.
+    ///
+    /// The default implementation appends an incorrect type diagnostic to `diagnostics`.
+    /// The expected type is retrieved from [Self::EXPECTED_TYPE].
+    fn visit_str(
+        self,
+        _value: TokenText,
+        range: TextRange,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self::Output> {
+        debug_assert!(
+            !Self::EXPECTED_TYPE.contains(ExpectedType::STR),
+            "This method should be implemented because the type is expected."
+        );
+        diagnostics.push(DeserializationDiagnostic::new_incorrect_type(
+            Self::EXPECTED_TYPE,
+            range,
+        ));
+        None
+    }
+
+    /// The visited value is an array-like (array, list, vector) structure.
+    ///
+    /// The default implementation appends an incorrect type diagnostic to `diagnostics`.
+    /// The expected type is retrieved from [Self::EXPECTED_TYPE].
+    fn visit_array(
+        self,
+        _items: impl Iterator<Item = impl DeserializableValue>,
+        range: TextRange,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self::Output> {
+        debug_assert!(
+            !Self::EXPECTED_TYPE.contains(ExpectedType::ARRAY),
+            "This method should be implemented because the type is expected."
+        );
+        diagnostics.push(DeserializationDiagnostic::new_incorrect_type(
+            Self::EXPECTED_TYPE,
+            range,
+        ));
+        None
+    }
+
+    /// The visited value is a `map` (key-value pairs).
+    ///
+    /// The default implementation appends an incorrect type diagnostic to `diagnostics`.
+    /// The expected type is retrieved from [Self::EXPECTED_TYPE].
+    fn visit_map(
+        self,
+        _members: impl Iterator<Item = (impl DeserializableValue, impl DeserializableValue)>,
+        range: TextRange,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<Self::Output> {
+        debug_assert!(
+            !Self::EXPECTED_TYPE.contains(ExpectedType::MAP),
+            "This method should be implemented because the type is expected."
+        );
+        diagnostics.push(DeserializationDiagnostic::new_incorrect_type(
+            Self::EXPECTED_TYPE,
+            range,
+        ));
+        None
+    }
+}
 
 /// A small type to interrogate the result of a JSON deserialization
 #[derive(Debug, Default)]
 pub struct Deserialized<P> {
     diagnostics: Vec<Error>,
-    deserialized: P,
+    deserialized: Option<P>,
 }
```

Comment:
Before, `P` was always guaranteed; now it's not. We should add a comment to `deserialized` and explain at least when/why it's an `Option` now.